File: MoonEngine-Root/Render/src/main/java/com/greentree/engine/moon/render/assets/material/PBRMaterialAssetSerializator.java
package com.greentree.engine.moon.render.assets.material;

import com.greentree.engine.moon.assets.asset.Asset;
import com.greentree.engine.moon.assets.asset.AssetKt;
import com.greentree.engine.moon.assets.asset.Value2Function;
import com.greentree.engine.moon.assets.key.AssetKey;
import com.greentree.engine.moon.assets.serializator.AssetSerializator;
import com.greentree.engine.moon.assets.serializator.manager.AssetManager;
import com.greentree.engine.moon.render.material.Material;
import com.greentree.engine.moon.render.material.MaterialProperties;
import com.greentree.engine.moon.render.shader.ShaderProgramData;

public final class PBRMaterialAssetSerializator implements AssetSerializator<Material> {

    @Override
    public Asset<Material> load(AssetManager manager, AssetKey key) {
        var program = manager.load(ShaderProgramData.class, "shader/pbr_mapping/pbr_mapping.glsl");
        var properties = manager.load(MaterialProperties.class, key);
        return AssetKt.map(program, properties, new PBRMaterialAssetSerializatorFunction());
    }

    private static final class PBRMaterialAssetSerializatorFunction
            implements Value2Function<ShaderProgramData, MaterialProperties, Material> {

        private static final long serialVersionUID = 1L;

        @Override
        public Material apply(ShaderProgramData program, MaterialProperties properties) {
            return new Material(program, properties);
        }

    }

}

File: MoonEngine-Root/Render/src/main/java/com/greentree/engine/moon/render/assets/material/PBRMaterialAssetSerializator.kt
package com.greentree.engine.moon.render.assets.material

import com.greentree.engine.moon.assets.asset.Asset
import com.greentree.engine.moon.assets.asset.Value2Function
import com.greentree.engine.moon.assets.asset.map
import com.greentree.engine.moon.assets.key.AssetKey
import com.greentree.engine.moon.assets.serializator.AssetSerializator
import com.greentree.engine.moon.assets.serializator.manager.AssetManager
import com.greentree.engine.moon.assets.serializator.manager.load
import com.greentree.engine.moon.render.material.Material
import com.greentree.engine.moon.render.material.MaterialProperties
import com.greentree.engine.moon.render.shader.ShaderProgramData

class PBRMaterialAssetSerializator : AssetSerializator<Material> {

	override fun load(manager: AssetManager, key: AssetKey): Asset<Material> {
		val program = manager.load<ShaderProgramData>("shader/pbr_mapping/pbr_mapping.glsl")
		val properties = manager.load<MaterialProperties>(key)
		return map(
			program,
			properties,
			PBRMaterialAssetSerializatorFunction
		)
	}

	private object PBRMaterialAssetSerializatorFunction :
		Value2Function<ShaderProgramData, MaterialProperties, Material> {

		override fun apply(program: ShaderProgramData, properties: MaterialProperties): Material {
			return Material(program, properties)
		}
	}
}